### PR TITLE
cisco ise asim network parser

### PR DESCRIPTION
**Pull request by spsocprime**
Required items, please complete
   
   Change(s):
   - Added ASIM Network Schema Parser for Cisco ISE
   - Added Sample Data to 'Sample Data/ASIM' folder

   Testing Completed:
   - Yes

### CODE REVIEW

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        disabled:bool=false){
+        let src_or_any=set_union(srcipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+        let dst_or_any=set_union(dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+        let ip_any = set_union(srcipaddr_has_any_prefix, dstipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+        Syslog
+        | where not(disabled)
+        | where ProcessName has_any ("CISE", "CSCO")
+        | parse kind = regex SyslogMessage with @"\d{10}\s" EventOriginalType @"\s(NOTICE|INFO|WARN|WARNING|ERROR|FATAL|DEBUG)"
+        //***************************** <Prefiltering> **************************
+        | where (isnull(starttime) or TimeGenerated>=starttime)
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:83]: Time filters can be applied even before ProcessName filter

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        let dst_or_any=set_union(dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+        let ip_any = set_union(srcipaddr_has_any_prefix, dstipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+        Syslog
+        | where not(disabled)
+        | where ProcessName has_any ("CISE", "CSCO")
+        | parse kind = regex SyslogMessage with @"\d{10}\s" EventOriginalType @"\s(NOTICE|INFO|WARN|WARNING|ERROR|FATAL|DEBUG)"
+        //***************************** <Prefiltering> **************************
+        | where (isnull(starttime) or TimeGenerated>=starttime) 
+            and (isnull(endtime) or TimeGenerated<=endtime) 
+            and not(disabled)
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:85]: Repeated

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
...ated with non-SSL error","LDAPS connection terminated with non-SSL error","90062","Failure","ERROR","Low","Cannot connect to Domain Controller","Cannot connect to Domain Controller","90063","Success","INFO","Informational","Successfully establish connection to Domain Controller","Successfully establish connection to Domain Controller","90066","Failure","ERROR","Low","Lost connection with Domain Controller","Lost connection with Domain Controller","90078","Success","INFO","Informational","Closed connection to Domain Controller","Closed connection to Domain Controller","91082","Failure","WARN","Low","RADIUS DTLS: Connection to OCSP server failed","RADIUS DTLS: Connection attempt to OCSP server failed.","11317","Failure","WARN","Low","TrustSec SSH connection failed","ISE failed to establish SSH connection to a network device. Verify network device SSH credentials in the Network Device page are similar to the credentials configured on the network device. Check network device enabled ssh connections from ISE (ip address)","5405","Failure","NOTICE","Low","RADIUS Request dropped","RADIUS request dropped","5406","Failure","NOTICE","Low","TACACS+ Request dropped","TACACS+ request dropped"];
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:53]: Based on the event, there should be DvcAction in Network Session as well.
For some of the initial records of LDAP, the events do not look like NetworkSession. They are more of Audit events.

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      | extend 
+              EventVendor = "Cisco"
+              , EventProduct = "ISE"
+              , EventProductVersion = "3.2"
+              , EventCount = int(1)
+              , EventSchema = "NetworkSession"
+              , EventSchemaVersion = "0.2.6"
+              , EventStartTime = coalesce(EventTime, TimeGenerated)
+              , EventEndTime = coalesce(EventTime, TimeGenerated)
+              , EventType = "NetworkSession"
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:109]: Can we check if EventSubType is applicable in some of these?

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 10 records (100.0%) for field [EventProduct] of type [Enumerated]: [""ISE""] (Schema:NetworkSession)"
+"(1) Warning: Empty value in 10 records (100.0%)  in mandatory field [Dvc] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [DstMacAddr] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [DstUsername] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [EventOriginalResultDetails] (Schema:NetworkSession)"
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv:6]: This value was coming from the lookup table. Why is this empty?

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 10 records (100.0%) for field [EventProduct] of type [Enumerated]: [""ISE""] (Schema:NetworkSession)"
+"(1) Warning: Empty value in 10 records (100.0%)  in mandatory field [Dvc] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [DstMacAddr] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [DstUsername] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [EventOriginalResultDetails] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in optional field [SrcMacAddr] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in recommended field [DvcIpAddr] (Schema:NetworkSession)"
+"(2) Info: Empty value in 10 records (100.0%)  in recommended field [SrcHostname] (Schema:NetworkSession)"
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv:9]: 10 records are too less for the amount of event types we are considering. Please test on more sample space.

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_SchemaTest.csv [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+"(2) Info: Missing optional field [ThreatFirstReportedTime]"
+"(2) Info: Missing optional field [ThreatId]"
+"(2) Info: Missing optional field [ThreatIpAddr]"
+"(2) Info: Missing optional field [ThreatIsActive]"
+"(2) Info: Missing optional field [ThreatLastReportedTime]"
+"(2) Info: Missing optional field [ThreatName]"
+"(2) Info: Missing optional field [ThreatOriginalConfidence]"
+"(2) Info: Missing optional field [ThreatOriginalRiskLevel]"
+"(2) Info: Missing optional field [ThreatRiskLevel]"
+"(2) Info: Missing recommended alias [Hostname] aliasing non-existent column [DstHostname]"
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_SchemaTest.csv:119]: Test results are missing for vim parser

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: added

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
...ated with non-SSL error","LDAPS connection terminated with non-SSL error","90062","Failure","ERROR","Low","Cannot connect to Domain Controller","Cannot connect to Domain Controller","90063","Success","INFO","Informational","Successfully establish connection to Domain Controller","Successfully establish connection to Domain Controller","90066","Failure","ERROR","Low","Lost connection with Domain Controller","Lost connection with Domain Controller","90078","Success","INFO","Informational","Closed connection to Domain Controller","Closed connection to Domain Controller","91082","Failure","WARN","Low","RADIUS DTLS: Connection to OCSP server failed","RADIUS DTLS: Connection attempt to OCSP server failed.","11317","Failure","WARN","Low","TrustSec SSH connection failed","ISE failed to establish SSH connection to a network device. Verify network device SSH credentials in the Network Device page are similar to the credentials configured on the network device. Check network device enabled ssh connections from ISE (ip address)","5405","Failure","NOTICE","Low","RADIUS Request dropped","RADIUS request dropped","5406","Failure","NOTICE","Low","TACACS+ Request dropped","TACACS+ request dropped"];
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml]: changed to `EventFieldsLookup`

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 1 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:82]: time filters moved before ProcessName filter

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 2 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: removed

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 3 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: Added DvcAction, lookup list updated

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv (hunk 5 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv]: added new test results

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_SchemaTest.csv (hunk 7 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_SchemaTest.csv]: added test results for vim parser

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv (hunk 6 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Tests/Cisco_ISE_NetworkSession_DataTest.csv]: New tests performed on larger dataset

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 4 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:108]: Added EventSubType to lookup table

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
...n Controller","90063","Success","Allow","","Start","INFO","Informational","Successfully establish connection to Domain Controller","Successfully establish connection to Domain Controller","90066","Failure","Drop","Terminated","End","ERROR","Low","Lost connection with Domain Controller","Lost connection with Domain Controller","90078","Success","Allow","","Start","INFO","Informational","Closed connection to Domain Controller","Closed connection to Domain Controller","91082","Failure","Drop","Terminated","End","WARN","Low","RADIUS DTLS: Connection to OCSP server failed","RADIUS DTLS: Connection attempt to OCSP server failed.","11317","Failure","Drop","Terminated","End","WARN","Low","TrustSec SSH connection failed","ISE failed to establish SSH connection to a network device. Verify network device SSH credentials in the Network Device page are similar to the credentials configured on the network device. Check network device enabled ssh connections from ISE (ip address)","5405","Failure","Drop","Terminated","End","NOTICE","Low","RADIUS Request dropped","RADIUS request dropped","5406","Failure","Drop","Terminated","End","NOTICE","Low","TACACS+ Request dropped","TACACS+ request dropped"];
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:53]: In my opinion, for 60234 the connection has disconnected. It is not a failure but a valid state. Please make the changes accordingly if this is true.

Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:53]: For 61026, it is secure shutdown. Is should be End in EventSubType

Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:53]: For 60509, EventResultDetails should be "Maximum Retry"
For 61231 and 61232, "Routing issue" is applicable.
For 24019 and 24030, "Unknown" is applicable.

Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:53]: All failure events cannot be "Drop" in DvcAction. Example: Handshake failure is a "Deny".
Please revisit this again.

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        let dst_or_any=set_union(dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+        let ip_any = set_union(srcipaddr_has_any_prefix, dstipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+        Syslog
+        | where not(disabled)
+        | where (isnull(starttime) or TimeGenerated>=starttime) 
+            and (isnull(endtime) or TimeGenerated<=endtime) 
+        | where ProcessName has_any ("CISE", "CSCO")
+        | parse kind = regex SyslogMessage with @"\d{10}\s" EventOriginalType @"\s(NOTICE|INFO|WARN|WARNING|ERROR|FATAL|DEBUG)"
+        | where  EventOriginalType in (EventOriginalTypeList)
+            and (array_length(dvcaction) == 0)
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:87]: Now we have DvcAction so this field is also valid. Please move this to EventORiginalTypeList as well.

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_ASimNetworkSession_SchemaTest.csv [hunk 11]
```diff
@@ -0,0 +1,116 @@
+﻿Result
+"(0) Error: Missing optional alias [User] aliasing existing column [DstUsername]"
```
Comment (devikamehra) [inline Parsers/ASimNetworkSession/Tests/Cisco_ISE_ASimNetworkSession_SchemaTest.csv:2]: Please resolve this error.

**Review by devikamehra — CHANGES_REQUESTED** [review]

File: Parsers/ASimNetworkSession/Tests/Cisco_ISE_ASimNetworkSession_SchemaTest.csv (hunk 11 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Tests/Cisco_ISE_ASimNetworkSession_SchemaTest.csv:2]: fixed

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 9 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: changed

Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: changed

Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: Considering that EventOriginalType for this event is ERROR then EventResult is Failure. But if you want we can change it

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+            and (array_length(ip_any) == 0 or has_any_ipv4_prefix(SyslogMessage,ip_any)) 
+            and (array_length(hostname_has_any) == 0 or SyslogMessage has_any(hostname_has_any)) 
+            and (isnull(dstportnumber) or SyslogMessage has (strcat('DestinationPort=', tostring(dstportnumber))))
+        | lookup EventFieldsLookup on EventOriginalType
+        | where (eventresult == "*" or eventresult == EventResult)
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: Applied as suggested

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 9 above)
Comment (devikamehra) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml]: I think in that case, leave it the way it is.

**Comment by v-atulyadav** [comment]
Hi @spsocprime, 
Thank you for these fixes. We will provide an update as soon as possible.

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml [hunk 13]
```diff
@@ -0,0 +1,81 @@
+Parser:
```
Comment (vakohl) [inline Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml:1]: Please update union ASIM and VIM parser files by adding this parser.

Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml:1]: updated

**Review by devikamehra — CHANGES_REQUESTED** [review]

**Comment by devikamehra** [comment]
@vakohl Except one change all the changes were completed. Please help resolve the conflicts in union parser.

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+        hostname_has_any:dynamic=dynamic([]), 
+        dvcaction:dynamic=dynamic([]), 
+        eventresult:string='*', 
+        disabled:bool=false){
+        let src_or_any=set_union(srcipaddr_has_any_prefix, ipaddr_has_any_prefix); 
+        let dst_or_any=set_union(dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+        let ip_any = set_union(srcipaddr_has_any_prefix, dstipaddr_has_any_prefix, ipaddr_has_any_prefix);
+        let EventOriginalTypeList = toscalar(EventFieldsLookup
+            | where (eventresult == "*" or eventresult == EventResult) 
+                    and (array_length(dvcaction) == 0)
```
[the remark this answers is not in the document]
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:80]: added suggested part to filtering section of the parser

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+  - Name: dvcaction
+    Type: dynamic
+    Default: dynamic([])
+  - Name: eventresult
+    Type: string
+    Default: '*'
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
```
Comment (vakohl) [inline Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:52]: @spsocprime Changes looks good. Please format query to re-organize code

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml (hunk 13 above)
Comment (vakohl) [inline Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml:1]: @spsocprime Changes looks good. Please format query to re-organize code

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 15 above)
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:52]: @spsocprime Can you please take a look?

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml (hunk 13 above)
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml:1]: @spsocprime Can you please take a look?

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml (hunk 15 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionCiscoISE.yaml:52]: formatted query

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml (hunk 13 above)
Comment (spsocprime) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionCiscoISE.yaml:1]: formatted query

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**